fix(uri): allows specifying port for base_url

stannum-l opened the pull request:
This patch set adds a new port parameter to .from_uri() when a base_url
is not specified but a port is, it will take the original URL and adjust
the port to the specified one.

Closes: #706

🚨Please review the [guidelines for contributing](https://github.com/schemathesis/schemathesis/blob/master/CONTRIBUTING.rst) to this repository.

### Description
Please explain the changes you made here.

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [ ] All tests passing
- [ ] Added a changelog entry
- [ ] Extended the README / documentation, if necessary

Stranger6667 commented on src/schemathesis/loaders.py:
```diff
@@ -54,13 +55,18 @@ def from_uri(
     kwargs.setdefault("headers", {}).setdefault("User-Agent", USER_AGENT)
+    if not base_url and port:
+        o = urlparse(uri)
+        loc = re.sub(r':(\d+)$', '', o.netloc)
```
The `ParseResult` class provides `hostname` and `port` - I suggest using them instead of regex (it will require catching `ValueError` in some corner case).

For example, an URL like this one will be mangled - "http://2001:0db8:85a3:0000:0000:8a2e:0370:7334/openapi.json".
The base URL will end up with a different IPv6 address. E.g. with port 8081 it will be "2001:0db8:85a3:0000:0000:8a2e:0370:8081" instead of "[2001:0db8:85a3:0000:0000:8a2e:0370:7334]:8081"

I also suggest adding handling & a test case for IPv6 addresses.

stannum-l replied:
Thanks - that's a good idea, didn't think of ipv6

Stranger6667 commented on src/schemathesis/loaders.py:
```diff
@@ -54,13 +55,18 @@ def from_uri(
     """Load a remote resource and parse to schema instance."""
     kwargs.setdefault("headers", {}).setdefault("User-Agent", USER_AGENT)
+    if not base_url and port:
+        o = urlparse(uri)
```
Let's use "parsed" as the variable name

stannum-l replied:
This will be removed in favor of using `yarl`.

Stranger6667 commented on test/test_loaders.py:
```diff
@@ -38,6 +40,13 @@ def test_base_url_override(openapi3_schema_url, url):
+    schema = schemathesis.from_uri(openapi3_schema_url, port=8080)
+    endpoint = next(schema.get_all_endpoints())
+    url = re.sub(r':(\d+)$', '', urlparse(openapi3_schema_url).netloc)
+    assert f"{url}:8080" in endpoint.base_url
```
For simplicity, I suggest `assert endpoint.base_url == "http://127.0.0.1:8081/schema.yaml"` (later I'll add a separate function to avoid hardcoding the address)

Stranger6667 reviewed (changes requested):
Thank you for your contribution :) 

I left a couple of comments - please, take a look :)

Also, please add a changelog entry

Stranger6667 commented:
As an idea - it might be easier to use [Yarl](https://yarl.readthedocs.io/en/latest/api.html#new-url-generation)

It seems like [there is already](https://yarl.readthedocs.io/en/latest/api.html#yarl.URL.with_port) a method we need:

```python
>>> URL('http://example.com:8888').with_port(9999)
URL('http://example.com:9999')
>>> URL('http://example.com:8888').with_port(None)
URL('http://example.com')
```

Stranger6667 commented:
And regarding IPv6 addresses that are not enclosed - let's raise an error in this case since [RFC2732](http://www.ietf.org/rfc/rfc2732.txt) implies square brackets

Stranger6667 commented on docs/changelog.rst:
```diff
@@ -6,6 +6,8 @@ Changelog
 `Unreleased`_
 -------------
 
+- Add ``port`` parameter to ``from_uri()`` method. `#706`_
```
Could you please add `.. _#706: https://github.com/schemathesis/schemathesis/issues/706` string to the end of this file where other links are placed? It will make the rendered changelog point to the proper issue location.

Stranger6667 commented on test/test_loaders.py:
```diff
@@ -1,6 +1,7 @@
 import json
 
 import pytest
+from yarl import URL
```
I believe we need to add `yarl` to direct dependencies via `poetry add yarl`

stannum-l replied:
Done

Stranger6667 reviewed (changes requested):
Looks awesome! Thank you! :) I left a couple of minor comments; please take a look. Also, you need to rebase - there are new entries in the changelog that conflict with this PR. You need to adjust it :)

Otherwise - we are good to go :)

Stranger6667 reviewed (approved):
Awesome! Thank you @stannum-l for implementing this! :)

Stranger6667 commented on test/test_loaders.py:
```diff
@@ -38,6 +39,22 @@ def test_base_url_override(openapi3_schema_url, url):
+    parts = list(map(int, url.host.split(".")))
+    ipv6_host = "2002:{:02x}{:02x}:{:02x}{:02x}::".format(*parts)
+    ipv6 = url.with_host("[%s]" % ipv6_host)
+    schema = schemathesis.from_uri(ipv6, port=8081)
```
Oh, I missed that the build is failing on this line (other failures come from a flaky test).

I think it could be enough to use mocking here instead of adapting the server. Something like

`mocked = mocker.patch("requests.request", wraps=simple_openapi)`

In the example above, the `simple_openapi` variable is the fixture defined in the root conftest file

Stranger6667 replied:
Oh, I missed that the build is failing on this line (other failures come from a flaky test).

I think it could be enough to use mocking here instead of adapting the server. Something like

`mocked = mocker.patch("requests.request", wraps=simple_openapi)`

In the example above, the `simple_openapi` variable is the fixture defined in the root conftest file

Stranger6667 commented:
Also, another small note on the failing "Verify commit messages" check - I plan to squash all commits into a single one before merging and will put the missing category there :)

stannum-l commented:
Thanks for the quick feedback. Fixed the title of the PR and squashed the commits, but the unit test is still failing due to a SchemaValidation exception in the `get_endpoint()`. Will take a look again later, checked this in for now.

stannum-l commented on test/test_loaders.py:
```diff
@@ -38,6 +40,27 @@ def test_base_url_override(openapi3_schema_url, url):
+    assert endpoint.base_url == "http://127.0.0.1:8081/schema.yaml"
+
+
+def test_port_override_with_ipv6(openapi3_schema_url, simple_openapi, mocker):
```
not sure if `simple_openapi` is the payload to mock in `requests.get` - it appears L60 will fail with

```
>           raise InvalidSchema("Schema parsing failed. Please check your schema.") from exc
E           schemathesis.exceptions.InvalidSchema: Schema parsing failed. Please check your schema.
```

Any suggestion, or I can go through the schema code in greater details.

Stranger6667 commented on test/test_loaders.py:
```diff
@@ -38,6 +40,27 @@ def test_base_url_override(openapi3_schema_url, url):
+    parts = list(map(int, url.host.split(".")))
+    ipv6_host = "2002:{:02x}{:02x}:{:02x}{:02x}::".format(*parts)
+    ipv6 = url.with_host("[%s]" % ipv6_host)
+    schema = schemathesis.from_uri(ipv6, validate_schema=False, port=8081)
```
It fails due to passing URL instance instead of a string. It should be str(ipv6) instead

Stranger6667 commented on pyproject.toml:
```diff
@@ -43,6 +43,8 @@ importlib_metadata = { version = "^1.1", python = "<3.8" }
 junit-xml = "^1.9"
 starlette = "^0.13.2"
+yarl = "^1.6.0"
+responses = "^0.12.0"
```
I don’t see this dependency used :) seems like it could be removed

stannum-l replied:
Thanks - fixed in next patch

Stranger6667 commented:
There was a PR merged recently and this one has conflicts - please rebase :)

Stranger6667 commented:
Amazing! Thank you very much @stannum-l ! :)
